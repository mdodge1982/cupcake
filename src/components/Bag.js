import React, {Component} from 'react';
import './Bag.css';

class Bag extends Component {
	constructor() {
		super();
		this.state = {
			handleKeyUp: null,
			handleKeyDown: null
		};
	}
	render() {
		const {facing,angle,xPos,width,height} = this.props;
		let className = 'Bag facing-'+facing;
		const rotation = -1*(angle-1.57);
		return (
			<div className={className}
				style={{
					left: xPos+'px',
					transform: 'rotate('+rotation+'rad)',
					width: width+'px',
					height: height+'px'
				}}></div>
		);
	}
	componentDidMount() {
		const handleKeyDown = (e) => this.handleKeyDown(e);
		const handleKeyUp = (e) => this.handleKeyUp(e);
		document.addEventListener('keydown',handleKeyDown);
		document.addEventListener('keyup',handleKeyUp);
		this.setState({
			handleKeyDown,
			handleKeyUp
		});
	}
	componentWillUnmount() {
		document.removeEventListener('keydown',this.state.handleKeyDown);
		document.removeEventListener('keyup',this.state.handleKeyUp);
	}
	handleKeyDown(e) {
		if([37,39].includes(e.keyCode)){
			const direction = e.keyCode===37 ? 'left' : 'right';
			this.props.onArrowDown(direction);
		}else if(e.keyCode===32){
			this.props.onSpacebarDown();
		}
	}
	handleKeyUp(e) {
		if([37,39].includes(e.keyCode)){
			this.props.onArrowUp();
		}
	}
}

export default Bag;
