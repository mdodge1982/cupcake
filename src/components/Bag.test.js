import React from 'react';
import ReactTestUtils from 'react-dom/test-utils';
import {Provider} from 'react-redux';
import {createStore} from 'redux';
import renderer from 'react-test-renderer';
import Bag from './Bag';

const bag = {
	facing: 'front',
	angle: 0,
	xPos: 145,
	width: 50,
	height: 100
};
const reducer = (state = {bag}) => state;
const store = createStore(reducer);

test('Initial render', () => {
	const component = renderer.create(
		<Provider store={store}>
			<Bag {...bag} />
		</Provider>
	);
	let tree = component.toJSON();
	expect(tree).toMatchSnapshot();
});

test('Render after moving', () => {
	bag.facing = 'left';
	bag.angle = 1.6;
	bag.xPos = 29;
	const component = renderer.create(
		<Provider store={store}>
			<Bag {...bag} />
		</Provider>
	);
	let tree = component.toJSON();
	expect(tree).toMatchSnapshot();
});


test('Keyboard events test', () => {
	const mockMove = jest.fn();
	const mockStop = jest.fn();
	const mockFrosting = jest.fn();
	const component = ReactTestUtils.renderIntoDocument(
		<Bag
			store={store}
			bag={bag}
			moveBag={mockMove}
			stopBag={mockStop}
			addFrosting={mockFrosting}
		/>
	);
	// const button = ReactTestUtils.findRenderedDOMComponentWithClass(component,'shoot');
	// ReactTestUtils.Simulate.keyDown(document, {keyCode: 37, which: 37});
	const keyDown32 = new KeyboardEvent('keydown',{
		keyCode: 32
	});
	const keyDown37 = new KeyboardEvent('keydown',{
		keyCode: 37
	});
	const keyDown39 = new KeyboardEvent('keydown',{
		keyCode: 39
	});
	const keyUp37 = new KeyboardEvent('keyup',{
		keyCode: 37
	});
	const keyUp39 = new KeyboardEvent('keyup',{
		keyCode: 39
	});
	document.dispatchEvent(keyDown32);
	document.dispatchEvent(keyDown39);
	document.dispatchEvent(keyUp39);
	document.dispatchEvent(keyDown37);
	document.dispatchEvent(keyUp37);
	expect(mockFrosting.mock.calls.length).toBe(1);
	expect(mockMove.mock.calls.length).toBe(2);
	expect(mockStop.mock.calls.length).toBe(2);
});
