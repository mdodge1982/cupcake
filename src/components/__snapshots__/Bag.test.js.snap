// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Initial render 1`] = `
<div
  className="Bag facing-front"
  onTouchEnd={[Function]}
  onTouchMove={[Function]}
  style={
    Object {
      "height": "100px",
      "left": "145px",
      "transform": "rotate(1.57rad)",
      "width": "50px",
    }
  }
/>
`;

exports[`Render after moving 1`] = `
<div
  className="Bag facing-left"
  onTouchEnd={[Function]}
  onTouchMove={[Function]}
  style={
    Object {
      "height": "100px",
      "left": "29px",
      "transform": "rotate(-0.030000000000000027rad)",
      "width": "50px",
    }
  }
/>
`;
